### DISCUSSION TITLE
TypeError: Cannot redefine property: Request         at Function.defineProperty (<anonymous>)

### DISCUSSION THREAD

**Discussion by halilibrahimcelik** [mswjs/msw#2288] [Q&A]
Okay after dealing with TextEncoder  and  ReadableStream is not defined errors, I finally got chance to test my api but now 
I'm getting  this error when I run my test  on my reservation.test.tsx file. 

    TypeError: Cannot redefine property: Request
        at Function.defineProperty (<anonymous>)

       9 | import { server } from './__tests__/__mocks__/msw/server';
      10 |
    > 11 | beforeAll(() => server.listen());
         |                        ^
      12 | afterEach(() => server.resetHandlers());
      13 | afterAll(() => server.close());
      14 |

      at recordRawFetchHeaders (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/interceptors/ClientRequest/utils/recordRawHeaders.ts:157:10)
      at _ClientRequestInterceptor.setup (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/interceptors/ClientRequest/index.ts:112:5)
      at _ClientRequestInterceptor.apply (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/Interceptor.ts:130:10)
      at BatchInterceptor.setup (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/BatchInterceptor.ts:44:19)
      at BatchInterceptor.apply (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/Interceptor.ts:130:10)
      at SetupServerApi.listen (node_modules/.pnpm/msw@2.4.7_typescript@4.5.4/node_modules/msw/src/node/SetupServerCommonApi.ts:102:22)
      at Object.listen (jest.setup.js:11:24)

https://github.com/halilibrahimcelik/jest-test-repo-base-venue this is  the project repo btw,  you can look at my configurations on msw  handlers and servers file as well as jest files.

Any help really appreciated.

**Comment by halilibrahimcelik** [comment]
When I look at   Discussed in #2271, I thought the problem might related to version, so  I changed  it from 2.4.8 to 2.4.7  as described but my problem was still there. So after further research, I saw this comment https://github.com/mswjs/msw/discussions/2271#discussioncomment-10613606   and finally 2.4.3 version solved my problem. 
 I also configured my setupfile like this 

> ```js
>   const { TextDecoder, TextEncoder } = require('node:util');
> const { ReadableStream } = require('node:stream/web');
 >Object.defineProperties(globalThis, {
 > TextDecoder: { value: TextDecoder },
 > TextEncoder: { value: TextEncoder },
  >ReadableStream: { value: ReadableStream },
>});
>
>const { Blob, File } = require('node:buffer');
>const { fetch, Headers, FormData, Request, Response } = require('undici');
>
>Object.defineProperties(globalThis, {
 > fetch: { value: fetch, writable: true },
>  Blob: { value: Blob },
  >File: { value: File },
 > Headers: { value: Headers },
 > FormData: { value: FormData },
>  Request: { value: Request },
  >Response: { value: Response },
>});
>
> ```

**Reply by Hagendorn** [reply]
I ran in the same issue and switched to this setup and removed `undici` completely:
https://github.com/mswjs/msw/issues/1916#issuecomment-1919965699
and with this switch the error is gone

**Reply by okusarobert** [reply]
> When I look at Discussed in #2271, I thought the problem might related to version, so I changed it from 2.4.8 to 2.4.7 as described but my problem was still there. So after further research, I saw this comment [#2271 (comment)](https://github.com/mswjs/msw/discussions/2271#discussioncomment-10613606) and finally 2.4.3 version solved my problem. I also configured my setupfile like this
> 
> > ```js
> >   const { TextDecoder, TextEncoder } = require('node:util');
> > const { ReadableStream } = require('node:stream/web');
> > Object.defineProperties(globalThis, {
> > TextDecoder: { value: TextDecoder },
> > TextEncoder: { value: TextEncoder },
> > ReadableStream: { value: ReadableStream },
> > });
> > 
> > const { Blob, File } = require('node:buffer');
> > const { fetch, Headers, FormData, Request, Response } = require('undici');
> > 
> > Object.defineProperties(globalThis, {
> > fetch: { value: fetch, writable: true },
> >  Blob: { value: Blob },
> > File: { value: File },
> > Headers: { value: Headers },
> > FormData: { value: FormData },
> >  Request: { value: Request },
> > Response: { value: Response },
> > });
> > ```

this solved my issue. Hopefully this gets fixed in the future versions

**Comment by iamlucasgomes** [comment]
> Okay after dealing with TextEncoder and ReadableStream is not defined errors, I finally got chance to test my api but now I'm getting this error when I run my test on my reservation.test.tsx file.
> 
> ```
> TypeError: Cannot redefine property: Request
>     at Function.defineProperty (<anonymous>)
> 
>    9 | import { server } from './__tests__/__mocks__/msw/server';
>   10 |
> > 11 | beforeAll(() => server.listen());
>      |                        ^
>   12 | afterEach(() => server.resetHandlers());
>   13 | afterAll(() => server.close());
>   14 |
> 
>   at recordRawFetchHeaders (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/interceptors/ClientRequest/utils/recordRawHeaders.ts:157:10)
>   at _ClientRequestInterceptor.setup (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/interceptors/ClientRequest/index.ts:112:5)
>   at _ClientRequestInterceptor.apply (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/Interceptor.ts:130:10)
>   at BatchInterceptor.setup (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/BatchInterceptor.ts:44:19)
>   at BatchInterceptor.apply (node_modules/.pnpm/@mswjs+interceptors@0.35.6/node_modules/@mswjs/interceptors/src/Interceptor.ts:130:10)
>   at SetupServerApi.listen (node_modules/.pnpm/msw@2.4.7_typescript@4.5.4/node_modules/msw/src/node/SetupServerCommonApi.ts:102:22)
>   at Object.listen (jest.setup.js:11:24)
> ```
> 
> https://github.com/halilibrahimcelik/jest-test-repo-base-venue this is the project repo btw, you can look at my configurations on msw handlers and servers file as well as jest files.
> 
> Any help really appreciated.



the error `
TypeError: Cannot redefine property: Request
    at Function.defineProperty (<anonymous>)
`

happens because in your polyfill, you didn't allow Request to be configurable

You can adjust it like this

```js
Object.defineProperties(globalThis, {
fetch: { value: fetch, writable: true },
 Blob: { value: Blob },
File: { value: File },
Headers: { value: Headers },
FormData: { value: FormData },
 Request: { value: Request, configurable: true },
Response: { value: Response, configurable: true },
});
```

**Reply by halilibrahimcelik** [reply]
Yes but also there is a version issue, when I lowered to 2.4.3  all my problems were gone away.

**Comment by takakikasuga** [comment]
[IMHO]

@halilibrahimcelik 

I refered this document and it's completely worked at msw@2.4.9

https://mswjs.io/docs/faq/#requestresponsetextencoder-is-not-defined-jest


```
// jest.polyfills.js
/**
 * @note The block below contains polyfills for Node.js globals
 * required for Jest to function when running JSDOM tests.
 * These HAVE to be require's and HAVE to be in this exact
 * order, since "undici" depends on the "TextEncoder" global API.
 *
 * Consider migrating to a more modern test runner if
 * you don't want to deal with this.
 */
 
const { TextDecoder, TextEncoder } = require('node:util')
 
Object.defineProperties(globalThis, {
  TextDecoder: { value: TextDecoder },
  TextEncoder: { value: TextEncoder },
})
 
const { Blob, File } = require('node:buffer')
const { fetch, Headers, FormData, Request, Response } = require('undici')
 
Object.defineProperties(globalThis, {
  fetch: { value: fetch, writable: true },
  Blob: { value: Blob },
  File: { value: File },
  Headers: { value: Headers },
  FormData: { value: FormData },
  Request: { value: Request, configurable: true },
  Response: { value: Response, configurable: true },
})
```

**Comment by kettanaito** [comment]
Please note that the [Jest usage instructions](https://mswjs.io/docs/migrations/1.x-to-2.x#requestresponsetextencoder-is-not-defined-jest) have been updated today. You are encouraged to use `jest-fixed-jsdom` environment instead of doing these manual patches.
